File: apps/ios/Packages/StoryArcKit/Sources/Persistence/StorageUsage.swift
public import Foundation

internal import WebKit

/// What StoryArc is using on disk, and how to give it back.
///
/// `settings-and-about` asks for cache, reading history and downloads to be "individually
/// clearable, each stating what it removes and how much space it frees". A number is the
/// point: "clear cache" with no size behind it asks a reader to guess whether it is worth
/// doing.
///
/// Downloads are not measured here. They are, but by whoever owns them: ``DownloadStore``
/// walks its own directory and the size is handed to the Privacy screen, because this type
/// knows about the cache and the history and has no business knowing where a download lands.
public struct StorageUsage: Sendable {

    /// The directory this measures and empties. The system's, unless a test says otherwise.
    private let caches: URL?

    /// The other half of what the row promises: the web view's cookies and origin storage.
    ///
    /// Injected so a test can watch it happen. The real one talks to `WKWebsiteDataStore`,
    /// which is a process-wide singleton — a test that called it would be clearing the test
    /// host's own website storage, which is not a decision a test gets to make.
    private let removeWebsiteData: @Sendable @MainActor () async -> Void

    public init(
        caches: URL? = nil,
        removeWebsiteData: @escaping @Sendable @MainActor () async -> Void
            = StorageUsage.removeAllWebsiteData
    ) {
        self.caches = caches
        self.removeWebsiteData = removeWebsiteData
    }

    /// Bytes the caches directory is holding. Includes the web view's own.
    public func cacheBytes() -> Int64 {
        guard let caches = cachesDirectory else { return 0 }
        return Self.size(of: caches)
    }

    /// Empties the caches directory, and the web view's cookies and origin storage with it.
    ///
    /// Contents rather than the directory itself: removing the directory out from under a
    /// web view that has it open is how the next page load finds nothing where it expected a
    /// writable path.
    ///
    /// The second half is what the row has always said and never did. "Decoded pages and
    /// web-view data" is the string on the Privacy screen in all four languages; the web
    /// view's *cache* does live in this directory, but its cookies and per-origin storage do
    /// not — so a publication that reached the network could leave an identifier behind that
    /// the reader had just been told was gone. Clearing them is the honest half of that
    /// choice, and cheaper than translating a smaller promise four times.
    ///
    /// Android's `StorageUsage.clearCache` clears `CookieManager` and `WebStorage` in the
    /// same call, for the same reason.
    public func clearCache() async {
        if let caches = cachesDirectory {
            let contents = (try? FileManager.default.contentsOfDirectory(
                at: caches, includingPropertiesForKeys: nil
            )) ?? []
            for item in contents {
                try? FileManager.default.removeItem(at: item)
            }
        }
        await removeWebsiteData()
    }

    /// Every kind of website data WebKit knows about, since the reader asked for all of it.
    ///
    /// `.distantPast` rather than a recent window: "clear" means clear.
    @MainActor
    public static func removeAllWebsiteData() async {
        await WKWebsiteDataStore.default().removeData(
            ofTypes: WKWebsiteDataStore.allWebsiteDataTypes(),
            modifiedSince: .distantPast
        )
    }

    private var cachesDirectory: URL? {
        caches ?? FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
    }

    private static func size(of directory: URL) -> Int64 {
        guard let walker = FileManager.default.enumerator(
            at: directory, includingPropertiesForKeys: [.fileSizeKey, .isRegularFileKey]
        ) else { return 0 }
        var total: Int64 = 0
        for case let url as URL in walker {
            let values = try? url.resourceValues(forKeys: [.fileSizeKey, .isRegularFileKey])
            guard values?.isRegularFile == true else { continue }
            total += Int64(values?.fileSize ?? 0)
        }
        return total
    }
}

/// A size a person can read, in their own locale.
///
/// Powers of 1024 with the SI names, which is what every file manager on both platforms
/// shows — matching the convention a reader already has beats being right about kibibytes.
///
/// The number is formatted through `Locale.current`, not by `String(format:)`. That
/// composes a fixed decimal point, so a French reader saw "1.4 MB" where every other app
/// on their phone says "1,4 Mo". `localization` requires "numbers, dates and file sizes"
/// to follow the locale, and a hand-composed float does not.
public func formattedBytes(_ bytes: Int64) -> String {
    switch bytes {
    case ..<1: "0 kB"
    case ..<1024: "1 kB"
    case ..<(1024 * 1024): "\(bytes / 1024) kB"
    case ..<(1024 * 1024 * 1024): scaled(bytes, by: 1024 * 1024, unit: "MB")
    default: scaled(bytes, by: 1024 * 1024 * 1024, unit: "GB")
    }
}

/// One decimal place, in the reader's own number format.
private func scaled(_ bytes: Int64, by divisor: Double, unit: String) -> String {
    let value = (Double(bytes) / divisor).formatted(
        .number.precision(.fractionLength(1)).locale(.current)
    )
    return "\(value) \(unit)"
}

File: apps/android/core/persistence/src/main/kotlin/app/storyarc/core/persistence/StorageUsage.kt
package app.storyarc.core.persistence

import android.content.Context
import android.webkit.CookieManager
import android.webkit.WebStorage
import java.io.File

/**
 * What StoryArc is using on disk, and how to give it back.
 *
 * `settings-and-about` asks for cache, reading history and downloads to be "individually
 * clearable, each stating what it removes and how much space it frees". A number is the
 * point: "clear cache" with no size behind it asks a reader to guess whether it is worth
 * doing.
 *
 * Downloads are not measured here. They are, but by whoever owns them: `DownloadStore`
 * walks its own directory and the size is handed to the Privacy screen, because this type
 * knows about the cache and the history and has no business knowing where a download lands.
 *
 * The primary constructor takes the places rather than the [Context] that names them, so
 * that a unit test can drive [clearCache] against a temporary directory and a lambda it can
 * watch. `CookieManager` and `WebStorage` are platform singletons backed by a WebView a JVM
 * test has none of, which is the same reason iOS injects its `WKWebsiteDataStore` call.
 */
class StorageUsage internal constructor(
    private val cacheDirectories: List<File>,
    private val progressDatabase: File,
    private val removeWebViewData: () -> Unit,
) {
    constructor(context: Context) : this(
        cacheDirectories = listOfNotNull(context.cacheDir, context.externalCacheDir),
        progressDatabase = context.getDatabasePath(PROGRESS_DATABASE),
        removeWebViewData = { clearWebViewData() },
    )

    /** Bytes the cache directories are holding. Includes the web view's own. */
    fun cacheBytes(): Long = cacheDirectories.sumOf { it.sizeOnDisk() }

    /** Bytes the reading-position database is holding, journals included. */
    fun historyBytes(): Long =
        File(progressDatabase.parent ?: return 0)
            .listFiles { file -> file.name.startsWith(progressDatabase.name) }
            ?.sumOf { it.length() }
            ?: 0

    /**
     * Empties the cache, and the web view's cookies and origin storage with it.
     *
     * Contents rather than the directory itself: deleting the directory out from under a
     * web view that has it open is how the next page load finds nothing where it expected a
     * writable path.
     *
     * The second half is what the row has always said and never did. "Decoded pages and
     * web-view data" is the string on the Privacy screen in all four languages; the web
     * view's *cache* does live in these directories, but its cookies and per-origin storage
     * do not. That matters more here than on iOS: Readium serves every publication from the
     * one origin `https://readium_package/`, so a cookie set by one book is not scoped to
     * that book -- it is a stable identifier across everything the reader opens, and it
     * survived the clear that told them it was gone.
     *
     * iOS's `StorageUsage.clearCache` removes `WKWebsiteDataStore`'s data in the same call,
     * for the same reason.
     */
    fun clearCache() {
        cacheDirectories.forEach { directory ->
            directory.listFiles()?.forEach { it.deleteRecursively() }
        }
        removeWebViewData()
    }

    private fun File.sizeOnDisk(): Long =
        walkBottomUp().filter { it.isFile }.sumOf { it.length() }

    private companion object {
        const val PROGRESS_DATABASE = "progress.db"

        /**
         * Cookies first, then origin storage, then a flush.
         *
         * `removeAllCookies` writes through asynchronously and nothing here waits on the
         * callback, but the flush is not optional: a process killed before the next
         * automatic flush would bring the cookies back.
         */
        fun clearWebViewData() {
            val cookies = CookieManager.getInstance()
            cookies.removeAllCookies(null)
            cookies.flush()
            WebStorage.getInstance().deleteAllData()
        }
    }
}
